state/protocol/badger/mutator.go

```diff
@@ -413,114 +416,137 @@ func (m *Mutator) sealExtend(candidate *flow.Block) (*flow.Seal, error) {
 	header := candidate.Header
 	payload := candidate.Payload
 
-	// map each seal to the block it is sealing for easy lookup; we will need to
-	// successfully connect _all_ of these seals to the last sealed block for
-	// the payload to be valid
-	byBlock := make(map[flow.Identifier]*flow.Seal)
-	for _, seal := range payload.Seals {
-		byBlock[seal.BlockID] = seal
-	}
-	if len(payload.Seals) != len(byBlock) {
-		return nil, state.NewInvalidExtensionErrorf("multiple seals for the same block")
-	}
-
-	// get the parent's block seal, which constitutes the beginning of the
+	// Get the latest seal in the fork that ends with block `ParentID`.
+	// This last sealed block in the fork constitutes the beginning of the
 	// sealing chain; if no seals are part of the payload, it will also be used
 	// for the candidate block, which remains at the same sealed state
 	last, err := m.state.seals.ByBlockID(header.ParentID)
 	if err != nil {
```

Comment:
Do you mean, why it would be an unexpected internal error if we didn't find the `latest seal` in the fork? 
I think the explanation would be
```golang
// Get the latest seal in the fork that ends with block `ParentID`.  The protocol state saves this 
// information for each block that has been successfully added the chain tree. Prior to entering this 
// method, we check in `headerExtend` that the candidate block's header is a valid extension of 
// the chain. Reaching this payload check here implies that `headerExtend` passed successfully, 
// i.e. the parent must already be part of the chain tree. Therefore, _not_ finding the latest sealed block
// in the fork up to the parent constitutes a fatal internal error. 
```

Does that sound like a convincing argument?

Reply:
>I think the explanation would be

Yeah, that's for the explanation. I wonder if makes sense to make such dependency more explicit by taking the seal from the input:

```
parentSeal, err := headerExtend(block)
err := sealExtend(block, parentSeal)
```

This way, `headerExtend` must be called before `sealExtend`, otherwise there is no `parentSeal`.

Reply:
I am a big proponent of making conceptual dependencies explicit in the code. Thanks for the suggestion Leo. 
+1